contribute/design/proposal-cstor-orchestration.md

```diff
@@ -0,0 +1,376 @@
+       
+     * create SP object
+       ```
+       apiVersion: openebs.io/v1alpha1
+       kind: StoragePool
+       metadata:
+         name: spc-7b99e406-1260-11e8-aa43-00505684eb2e
+         labels:
+           "kubernetes.io/hostname": "node-host-label"
+       spec:
+         type: openebs-cstor
+         disks:
+           #Disks that are associated with this Pool.
+           disk-list: ["disk-name1", "disk-name2",...]
+           #This can contain the list of Disks that can be used as hot-spares if any of the
+           # used disk are errored.
+           #disk-list-spares: ["disk-namex", "disk-namey",...]
+         poolspec: 
+           #Defines the type of pool as passed from the SPC. stripe or mirror. 
+           pooltype: "stripe"
+           #Save the resources as sent from SPC. If these values are updated, 
+           # the corresponding values on the CStorPool Deployment YAML will have to be updated.
+           #resources:
+             #cpu: 
+             #memory: 
+           #Pool features as passed from the SPC.
+           #overprovisioning: false       
+       ```
+   * Admin associates the SPC with a StorageClass
+     ```
+     apiVersion: storage.k8s.io/v1
+     kind: StorageClass
+     metadata:
+       name: openebs-pool1
+     provisioner: openebs.io/provisioner-iscsi
+     parameters:
+       openebs.io/storage-pool-claim: pool1
+     ```
+
+2. Creating Volume using the (cStor) Storage Pools:
+
+   * Admin will create a PVC that is associated to StoragePool (linked by the pool name). 
+     ```
+     apiVersion: v1
+     kind: PersistentVolumeClaim
+     metadata:
+       name: demo-vol
+     spec:
+       accessModes:
+         - ReadWriteOnce
+       resources:
+         requests:
+           storage: 5G
+       storageClassName: openebs-pool1
+     ```
+
+   * maya-apiserver will create the OV cStorService (to get the portal ip address) and the OV cStorController Deployment that will contain cstor-ctrl (iscsi target) container and a cstor-ctrl-mgmt side-car. The configuration options for running the cstor-ctrl(iscsi target) will be passed via CStorVolume CR. 
+     - The cStorService YAML will have the following details:
+       ```
+       apiVersion: v1
+       kind: Service
+       metadata:
+         name: pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-ctrl-service
+         labels:
+           pv: pvc-ee171da3-07d5-11e8-a5be-42010a8001be
+       spec:
+         type: ClusterIP
+         ports:
+         - name: iscsi
+           port: 3260
+           protocol: TCP
+           targetPort: 3260
+         selector:
+           openebs/controller: cstor-controller
+           pv: pvc-ee171da3-07d5-11e8-a5be-42010a8001be
+       ```
+       
+     - The CStorVolume CR will contain the following details:
+       ```
+       apiVersion: openebs.io/v1alpha1
+       kind: CStorVolume
+       metadata:
+         name: pvc-ee171da3-07d5-11e8-a5be-42010a8001be-cstor-volume
+       spec:
```

Comment:
Hope I'm not too late for this.
How are we handling the case of adding new replica?
controller need to know the existing replicas vs new replicas, and also, it need to add new replicas to existing set after some conditions like rebuild is done/fetching status from replica etc.

Reply:
never too late @vishnuitta 

So the CStorVolume CR doesn't have any information about the replica. Instead, the CStorVolumeReplica CR has a reference to the CStorVolume (IP Address). Will the following work:

- CStorVolumeReplica will invoke a API on CStorVolume (controller), to register itself as a new replica - passing the self IP address and ID?
- CStorVolume - will set the state as new replica, kick-start a resync/rebuild with already existing replicas. CStorVolume will use the IP address passed int he registration to call the API on the CStorVolumeReplica for state transitions and checking status.

Reply:
yes @kmova .. this approach works. I think - we need to have workflow for adding new node.
Probably, CRD should have status field.
Or this workflow for addition of new node can be post MVP.

Reply:
So, few things to go into workflow are:
- what if replica/controller restarts with sync still going on?
- when do we say that replica i.e., new node is successfully added?

Reply:
thanks @vishnuitta added this info to design doc